### ISSUE TITLE
Consider FFI

### ISSUE THREAD

**Issue by blamario** [Plutonomicon/plutarch#134]
It would be a very nice feature if Plutarch could be used to rewrite only critical parts of on-chain code. The medium-term future looks like this:

1. as the block size grows, so will Plutus scripts,
2. as the scripts grow so will the number of functions and data types,
3. more and more of which will be shared with the off-chain code.

If a user needs to optimize one of those complex scripts, the prospect of rewriting *everything* in Plutarch may be daunting. Furthermore, the loss of code sharing between the on-chain and off-chain logic would be detrimental to maintenance.

To help out with this we should provide a sort of Plutus-Plutarch FFI, with the ability to call a Plutarch function from Plutus and vice versa, with (hopefully ergonomic and minimal) marshalling of arguments and results.

**Comment by L-as** [comment]
I don't think this is possible without work on the PlutusTx side unfortunately.

**Comment by TotallyNotChase** [comment]
Wasn't this sort of the point behind `PConstant` and `PLift`? I think they should allow much of the "FFI" envisioned. Though it would be limited to normal data types.

**Comment by blamario** [comment]
At least *some* of it should be possible today. Assuming I define a Haskell function `factorial :: Integer -> Integer` (stupid example, but bear with me) and compile it with `coreFactorial = $$(PlutusTx.compile [|| fac ||])`, I should be able to say something like

```
plutarchFactorial :: Term s (PInteger :--> PInteger)
plutarchFactorial = Plutarch.importCore coreFactorial
```

**Comment by L-as** [comment]
Sure, you can add another constructor to `RawTerm` for hoisted (and thus closed) UPLC terms.

**Comment by L-as** [comment]
Then we could have `embedUplc`

**Comment by blamario** [comment]
We'd lose the type `a` of `PlutusTx.CompiledCode a` that way though.

**Comment by blamario** [comment]
What I meant is that

    unsafeEmbedUplc :: CompiledCode a -> Term s b

is less safe than something like

    lessUnsafeEmbedUplc :: (PlutusType a, PInner a b ~ a) => CompiledCode b -> Term s a

though I think we'd need a new type family instead. But yes, we can still store it as a `RawTerm` in the end.

**Comment by L-as** [comment]
I wonder whether we could use `PConstant a` for `CompiledCode a`.

**Comment by blamario** [comment]
That's why I mentioned marshalling.

**Comment by blamario** [comment]
Fixed by PR #339, originally PR #264.
